src/data-model-providers/codegen/ServerClusterInterfaceRegistry.cpp

```diff
@@ -158,5 +178,59 @@ ServerClusterInterface * ServerClusterInterfaceRegistry::Get(const ConcreteClust
     return nullptr;
 }
 
+CHIP_ERROR ServerClusterInterfaceRegistry::SetContext(const ServerClusterContext & context)
+{
+    if (mContextIsValid)
+    {
+        // if there is no difference, do not re-initialize.
+        VerifyOrReturnError(mContext != context, CHIP_NO_ERROR);
+        ClearContext();
+    }
+
+    mContext         = context;
+    mContextIsValid  = true;
+    bool had_failure = false;
+
+    for (ServerClusterRegistration * registration = mRegistrations; registration != nullptr; registration = registration->next)
+    {
+        CHIP_ERROR err = registration->serverClusterInterface->Startup(&mContext);
+        if (err != CHIP_NO_ERROR)
+        {
+#if CHIP_ERROR_LOGGING
+            const ConcreteClusterPath path = registration->serverClusterInterface->GetPath();
+            ChipLogError(DataManagement, "Cluster %u/" ChipLogFormatMEI " startup failed: %" CHIP_ERROR_FORMAT, path.mEndpointId,
+                         ChipLogValueMEI(path.mClusterId), err.Format());
+#endif
+            had_failure = true;
+            // NOTE: this makes the object be in an awkward state:
+            //       - cluster is not initialized
+            //       - mContext is valid
+            //       As a result, ::Shutdown on this cluster WILL be called even if startup failed.
```

Comment:
Raising the obvious question: do we need this context mutator at all?  I'd think the context should be set up once, at registry init.

Reply:
It should be, however the way current code is structured we have our init ordering very "organic" (i.e. people added serverinit in whatever way they could until it worked).

What should happen is:
  - registry initialized first (and clusters should not be able to be added until that is done)
  - clusters initialized next

I can also try that approach, however I am unsure if I get a more difficult time integrating things if I go that path.

Reply:
Followup fodder: The registry should be initialized by the initializer of the provider, no?

Reply:
Yes, the call sequence today is:
  - `CodegenDataModelProvider::Startup` -> `mRegistry::SetContext` -> `ServerClusterInterface::Startup()`
 - `CodegenDatamodelProvider::SHutdown` -> `mRegistry::ClearContext` -> `ServerClusterInterface::Shutdown()`